src/components/problem-layout/ProblemCard.js

```diff
@@ -382,96 +405,233 @@ class ProblemCard extends React.Component {
+                                ),
+                            });
+                        }
+                        return null;
+                    });
+                }
+            
+                this.setState({
+                    hints: this.hints,
+                    giveDynamicHint: false, // Switch to manual hints
+                    activeHintType: "normal",
+                    dynamicHint: "Failed to generate AI hint. Displaying manual hints.",
+                    hintsFinished: new Array(this.hints.length).fill(0),
                 });
-                this.context.firebase.log(
-                    parsed,
-                    this.props.problemID,
-                    this.step,
-                    "",
-                    isCorrect,
-                    this.state.hintsFinished,
-                    "requestDynamicHint",
-                    chooseVariables(
-                        Object.assign(
-                            {},
-                            this.props.problemVars,
-                            this.props.variabilization
-                        ),
-                        this.props.seed
+            };            
+        
+            // Call ChatGPT to fetch the dynamic hint using streaming
+            fetchDynamicHint(
+                DYNAMIC_HINT_URL,
+                this.generateGPTHintParameters(this.prompt_template, this.state.bioInfo),
+                onChunkReceived,
+                onSuccessfulCompletion,
+                onError,
+                this.props.problemID,
+                chooseVariables(
+                    Object.assign(
+                        {},
+                        this.props.problemVars,
+                        this.step.variabilization
                     ),
-                    this.props.lesson,
-                    this.props.courseName,
-                    "dynamic",
-                    this.state.dynamicHint,
-                    this.state.bioInfo
-                );
-            })
-            .catch((error) => {
-                console.error(error);
-            });
-    };
+                    this.props.seed
+                ),
+                this.context
+            );
+        
+            this.context.firebase.log(
+                parsed,
+                this.props.problemID,
+                this.step,
+                "",
+                isCorrect,
+                this.state.hintsFinished,
+                "requestDynamicHint",
+                chooseVariables(
+                    Object.assign(
+                        {},
+                        this.props.problemVars,
+                        this.props.variabilization
+                    ),
+                    this.props.seed
+                ),
+                this.props.lesson,
+                this.props.courseName,
+                "dynamic",
+                this.state.dynamicHint,
+                this.state.bioInfo
+            );
+        };
+        
```

Comment:
Quick question: are you logging 1) when the dynamic hint is opened and 2) when the regenerate button is clicked?

Reply:
I added a TODO comment so we can do this in another PR